packages/engine/paima-sm/src/index.ts

```diff
@@ -383,88 +422,266 @@ async function processScheduledData(
+          input_data: inputData.inputData,
+          user_address: inputData.userAddress,
         },
         DBConn
       );
-      if (ENV.STORE_HISTORICAL_GAME_INPUTS) {
-        await storeGameInput.run(
-          {
-            block_height: latestChainData.blockNumber,
-            input_data: inputData.inputData,
-            user_address: inputData.userAddress,
-          },
-          DBConn
-        );
+    }
+  } catch (err) {
+    doLog(`[paima-sm] Database error on gameStateTransition: ${err}`);
+    throw err;
+  }
+}
+
+// Extract execution hints that define prerequisites for the execution.
+// @x| -> execution hint for the address the sender
+// z|*abc -> execution hint is the literal key 'abc'
+// NOTE: @x and *wallet can match as the same address
+function extractKeys(submittedData: SubmittedData): string[] {
+  const keys = [];
+  const args = submittedData.inputData.split('|');
+  const header = args.shift() as string;
+  if (header.match(/^@/)) {
+    keys.push(submittedData.realAddress.toLocaleLowerCase());
+  }
+  args.forEach(arg => {
+    if (arg.match(/^\*/)) {
+      const key = arg.replace(/^\*/, '').toLocaleLowerCase();
+      if (key) keys.push(key);
+    }
+  });
+  return keys;
+}
+
+type ExecutionNode = {
+  parents: ExecutionNode[];
+  keys: string[];
+  input: SubmittedData | null;
+  children: ExecutionNode[];
+  executed: boolean;
+};
+type BlockingLeaf = { key: string; node: ExecutionNode };
+
+/* Generate a serial tree (list) of execution nodes based on the submitted data. */
+function generateSerialExecutionTree(latestChainData: ChainData): ExecutionNode {
+  const treeRoot: ExecutionNode = {
+    parents: [],
+    keys: [],
+    input: null, // root has not input
+    children: [],
+    executed: false,
+  };
+  let currentNode = treeRoot;
+  for (const submittedData of latestChainData.submittedData) {
+    const node: ExecutionNode = {
+      keys: extractKeys(submittedData),
+      input: submittedData,
+      children: [],
+      parents: [currentNode],
+      executed: false,
+    };
+    currentNode.children = [node];
+    currentNode = node;
+  }
+  return treeRoot;
+}
+
+/* Generate a parallel tree of execution nodes based on the submitted data. */
+function generateParallelExecutionTree(latestChainData: ChainData): ExecutionNode {
+  const treeRoot: ExecutionNode = {
+    parents: [],
+    keys: [],
+    input: null, // root has not input
+    children: [],
+    executed: false,
+  };
+
+  const treeEndNodes: BlockingLeaf[] = [];
+  for (const submittedData of latestChainData.submittedData) {
+    const keys = extractKeys(submittedData);
+    const node: ExecutionNode = {
+      keys,
+      input: submittedData,
+      children: [],
+      parents: [],
+      executed: false,
+    };
+
+    if (!keys.length) {
+      node.parents = [treeRoot];
+      treeRoot.children.push(node);
+    } else {
+      const blocked = node.keys
+        .map(k => treeEndNodes.find(e => e.key === k))
```

Comment:
I don't think this is going to work. The point of the parallel notation is that all entries that use the same identifier all get run sequentially

It looks like what you have implemented instead is that the first entry with this identifier gets executed, and then every other input that uses that identifier gets executed in parallel. Also, your implementation is O(n^2) in the number of inputs to process which isn't ideal

I think probably it's better to use a library like https://www.npmjs.com/package/dependency-graph 

1. `addNode` for every `submittedData` where the node name is just the index
2. Keep a mapping `<key, largest index containing that key>`
3. For every submittedData, find all its keys, and `addDependency` to the latest value in the mapping of (2)
Now, when executing, do
1. `nextNodes = overallOrder(true)`
2. `await nextNodes`
3. `nextNodes.forEach(node => removeNode(node))`

This should have better performance and be easier to read than managing our own execution node system

Reply:
Also note there is an issue with the current implementation of the STF where a single error thrown by an STF call drops all transactions in a block. Ideally this would be changed so that failures only drop the submittedData that caused errors (easy change: just move the `SAVEPOINT` before around for loop for the `sqlQueries`)

The reason I bring this up is because it's less clear how this savepoint system works when parallelism mode is turned on. Most likely it would have to be something like this:
1. Run all the STF in parallel to get the list of queries to executive (if any of the STF failed, just ignore it)
2. Execute all the SQL queries in parallel after all the STF are done
3. If any of the SQL queries fails, exclude all sql queries generated from that `submittedData` and try again

However it's not immediately clear to me know we would implement (3) since I'm not sure how we can know which SQL query specifically was the cause of the crash and how we can know which `submittedData` it came from (recall: if one query crashes, all other queries that happen after it will crash I think with an error saying the transaction is in a bad state)